src/context.cpp

```diff
@@ -483,6 +484,102 @@ namespace librealsense
 
+
+    // Frame data and metadata are sent as two seperate streams.
+    // They are synchronized using frame_id for matching data+metadata sets.
+    template< class T >
+    class frame_metadata_syncer
+    {
+    public:
+        // We don't want the queue to get large, it means lots of drops and data that we store to (probably) throw later
+        static constexpr size_t max_md_queue_size = 8;
+        // If a metadata is lost we wait for it until the next frame arrives, causing a small delay but we prefer passing
+        // the frame late and without metadata over losing it.
+        static constexpr size_t max_frame_queue_size = 2;
+
+        void enqueue_frame( T && frame, std::shared_ptr< rs2_stream_profile > prof )
+        {
+            std::lock_guard< std::mutex > lock( _queues_lock );
+            _frame_queue.push_back( std::move( frame ) );
+            _prof_queue.push_back( prof );
+            search_for_match(); //Call under lock
+        }
+
+        void enqueue_metadata( realdds::topics::flexible_msg && md )
+        {
+            std::lock_guard< std::mutex > lock( _queues_lock );
+            while( _metadata_queue.size() >= max_md_queue_size )
+                _metadata_queue.pop_front(); //throw oldest
+
+            _metadata_queue.push_back( std::move( md ) );
+            search_for_match(); //Call under lock
+        }
+
+        typedef std::function< void( T && ) > on_frame_ready;
+        typedef std::function< void( T & frame, const realdds::topics::flexible_msg & metadata ) > add_metadata_to_frame;
+
+        void reset( on_frame_ready cb = nullptr, add_metadata_to_frame add_md = nullptr )
+        {
+            std::lock_guard< std::mutex > lock( _queues_lock );
+            _frame_queue.clear();
+            _prof_queue.clear();
+            _metadata_queue.clear();
+            _cb = cb;
+            _add_md = add_md;
+        }
+
+    private:
+        //Call under lock
+        void search_for_match()
+        {
+            // Wait for frame + metadata set, but if metadata is lost pass frame to the user anyway
+            if( _frame_queue.empty() || ( _metadata_queue.empty() && _frame_queue.size() < max_frame_queue_size ) )
+                return;
+
+            // Sync using frame ID. Frame IDs are assumed to be increasing with time
+            int frame_id = _frame_queue.front().frame_number;
+            int md_id = _metadata_queue.empty() ? frame_id : // If no metadata create a "match"
+                        std::stoi( rsutils::json::get< std::string >( _metadata_queue.front().json_data(), "frame-id" ) );
+
+            while( frame_id > md_id && _metadata_queue.size() > 1 )
+            {
+                // Metadata without frame, remove it from queue and check next
+                _metadata_queue.pop_front();
+                md_id = std::stoi( rsutils::json::get< std::string >( _metadata_queue.front().json_data(), "frame-id" ) );
+            }
+
+            if( frame_id == md_id )
```

Comment:
What happens if the frame-id is < than the metadata-id?
Doesn't look like it's handled?

Reply:
Fixed